### PR TITLE
feat: add exclusion configs for load interface

### PR DESCRIPTION
**Pull request by wildeslam**

### CODE REVIEW

File: arex-storage-config/src/main/java/com/arextest/config/model/vo/AgentRemoteConfigurationResponse.java [hunk 1]
```diff
@@ -35,4 +36,8 @@ public class AgentRemoteConfigurationResponse {
   private Map<String, String> extendField;
```
Comment (lucas-myx) [inline arex-storage-config/src/main/java/com/arextest/config/model/vo/AgentRemoteConfigurationResponse.java:37]:
```
private CompareConfiguration compareConfiguration;

public static class CompareConfiguration {
        private List<ConfigComparisonExclusions> comparisonExclusions;
        private Set<List<String>> globalExclusionList;
        private Set<String> ignoreNodeSet;
}
```

Comment (wildeslam) [reply arex-storage-config/src/main/java/com/arextest/config/model/vo/AgentRemoteConfigurationResponse.java:36]: fixed

File: arex-storage-web-api/src/main/java/com/arextest/storage/service/QueryConfigService.java [hunk 2]
```diff
@@ -75,6 +106,110 @@ public ScheduleReplayConfigurationResponse queryScheduleReplayConfiguration(Stri
         ScheduleReplayConfigurationResponse.class);
   }
 
+  public CompareConfiguration queryCompareConfiguration(String appId) {
+    CompareConfiguration compareConfiguration = new CompareConfiguration();
+
+    List<ConfigComparisonExclusions> comparisonExclusions = queryComparisonExclusions(appId);
+    for (ConfigComparisonExclusions exclusion : comparisonExclusions) {
+      if (exclusion.getCategoryType() == null && exclusion.getOperationName() == null) {
+        comparisonExclusions.remove(exclusion);
```
Comment (lucas-myx) [inline arex-storage-web-api/src/main/java/com/arextest/storage/service/QueryConfigService.java:115]: 直接在循环中删除List的元素可能会导致ConcurrentModificationException，可以使用Iterator的remove方法，或者List的removeIf方法

File: arex-storage-web-api/src/main/java/com/arextest/storage/web/controller/config/AgentRemoteConfigurationController.java [hunk 3]
```diff
@@ -147,13 +145,9 @@ public Response load(@RequestBody AgentRemoteConfigurationRequest request) {
 
       // asynchronously update application env
       asyncUpdateAppEnv(requestInstance);
+      CompareConfiguration compareConfiguration = queryConfigService.queryCompareConfiguration(appId);
```
Comment (lucas-myx) [inline arex-storage-web-api/src/main/java/com/arextest/storage/web/controller/config/AgentRemoteConfigurationController.java:148]: suggest `queryCompareConfiguration` method add try catch, avoid throw exception to affect normal logic.

File: arex-storage-web-api/src/main/java/com/arextest/storage/service/QueryConfigService.java [hunk 4]
```diff
@@ -75,6 +106,110 @@ public ScheduleReplayConfigurationResponse queryScheduleReplayConfiguration(Stri
...
+        .collect(Collectors.groupingBy(
+            config -> config.getOperationName() + config.getOperationType()));
+    List<ConfigComparisonExclusions> result = new ArrayList<>();
+    for (Entry<String, List<ComparisonExclusionsConfiguration>> entry : groupByOperationNameAndType
+        .entrySet()) {
+      List<ComparisonExclusionsConfiguration> configList = entry.getValue();
+      if (CollectionUtils.isEmpty(configList)) {
+        continue;
+      }
+      ConfigComparisonExclusions vo = new ConfigComparisonExclusions();
```
Comment (lucas-myx) [inline arex-storage-web-api/src/main/java/com/arextest/storage/service/QueryConfigService.java:203]: Try not to name it with `vo` as much as possible, replace `configComparisonExclusions`

File: arex-storage-config/src/main/java/com/arextest/config/model/dao/config/AppContractCollection.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+/**
+ * @author wildeslam.
+ * @create 2024/7/4 10:57
+ */
+@Data
+@EqualsAndHashCode(callSuper = true)
+@FieldNameConstants
+@Document(collection = "AppContract")
+public class AppContractCollection extends BaseEntity {
```
Comment (coryhh) [inline arex-storage-config/src/main/java/com/arextest/config/model/dao/config/AppContractCollection.java:17]: The arex-storage-config package is referenced by arex-api. If the model and logic is not shared, it is recommended to include it in arex-storage-core.

File: arex-storage-web-api/src/main/java/com/arextest/storage/service/QueryConfigService.java [hunk 6]
```diff
@@ -75,6 +107,118 @@ public ScheduleReplayConfigurationResponse queryScheduleReplayConfiguration(Stri
...
+        } else if (operationConfiguration.getOperationTypes().size() == 1) {
+          String operationType = operationConfiguration.getOperationTypes().iterator().next();
+          config.setOperationType(operationType);
+          newConfigs.add(config);
+        }
+      } else {
+        newConfigs.add(config);
+      }
+    }
+
```
Comment (lucas-myx) [inline arex-storage-web-api/src/main/java/com/arextest/storage/service/QueryConfigService.java:199]: suggest split independent methods for easy reading

File: arex-storage-web-api/src/main/java/com/arextest/storage/service/QueryConfigService.java [hunk 7]
```diff
@@ -75,6 +107,118 @@ public ScheduleReplayConfigurationResponse queryScheduleReplayConfiguration(Stri
...
+    Map<String, List<ComparisonExclusionsConfiguration>> groupByOperationNameAndType = newConfigs
+        .stream()
+        .collect(
+            Collectors.groupingBy(config -> config.getOperationName() + config.getOperationType()));
+    List<ConfigComparisonExclusions> result = new ArrayList<>();
+    for (Entry<String, List<ComparisonExclusionsConfiguration>> entry : groupByOperationNameAndType
+        .entrySet()) {
+      List<ComparisonExclusionsConfiguration> configList = entry.getValue();
```
Comment (lucas-myx) [inline arex-storage-web-api/src/main/java/com/arextest/storage/service/QueryConfigService.java:207]: `for (List<ComparisonExclusionsConfiguration> configList : groupByOperationNameAndType.values())`

File: arex-storage-config/src/main/java/com/arextest/config/model/vo/ConfigComparisonExclusions.java [hunk 8]
```diff
@@ -10,10 +11,8 @@
  */
 @Data
 @NoArgsConstructor
-public class ConfigComparisonExclusionsVO {
-  private String appId;
-  private String operationId;
-  private String dependencyId;
-  private String operationType;
-  private List<String> exclusions;
+public class ConfigComparisonExclusions {
```
Comment (coryhh) [inline arex-storage-config/src/main/java/com/arextest/config/model/vo/ConfigComparisonExclusions.java:14]: improve the name of Class, the name of Class should be named in the singular form if possible.

File: arex-storage-web-api/src/main/java/com/arextest/storage/service/QueryConfigService.java (hunk 7 above)
Comment (wildeslam) [reply arex-storage-web-api/src/main/java/com/arextest/storage/service/QueryConfigService.java]: fixed

File: arex-storage-web-api/src/main/java/com/arextest/storage/service/QueryConfigService.java [hunk 9]
```diff
@@ -75,6 +108,151 @@ public ScheduleReplayConfigurationResponse queryScheduleReplayConfiguration(Stri
...
+
+    compareConfiguration.setComparisonExclusions(comparisonExclusions);
+    compareConfiguration.setIgnoreNodeSet(getIgnoreNodeSet());
+
+    return compareConfiguration;
+  }
+
+  private List<ConfigComparisonExclusion> queryComparisonExclusions(String appId) {
+    // ConfigComparisonExclusion, Source data
+    List<ComparisonExclusionsConfiguration> configs = comparisonExclusionsConfigurationRepository.listBy(
```
Comment (coryhh) [inline arex-storage-web-api/src/main/java/com/arextest/storage/service/QueryConfigService.java:133]: filter expired config

File: arex-storage-web-api/src/main/java/com/arextest/storage/service/QueryConfigService.java [hunk 10]
```diff
@@ -75,6 +108,151 @@ public ScheduleReplayConfigurationResponse queryScheduleReplayConfiguration(Stri
...
+        if (operationConfiguration.getOperationTypes().size() > 1) {
+          for (String operationType : operationConfiguration.getOperationTypes()) {
+            ComparisonExclusionsConfiguration newConfig = new ComparisonExclusionsConfiguration();
+            newConfig.setOperationName(config.getOperationName());
+            newConfig.setOperationType(operationType);
+            newConfig.setExclusions(config.getExclusions());
+            newConfigs.add(newConfig);
+          }
+        } else if (operationConfiguration.getOperationTypes().size() == 1) {
+          String operationType = operationConfiguration.getOperationTypes().iterator().next();
```
Comment (coryhh) [inline arex-storage-web-api/src/main/java/com/arextest/storage/service/QueryConfigService.java:247]: Why generate a new iterator object

File: arex-storage-web-api/src/main/java/com/arextest/storage/web/controller/config/AgentRemoteConfigurationController.java [hunk 11]
```diff
@@ -147,13 +145,14 @@ public Response load(@RequestBody AgentRemoteConfigurationRequest request) {
...
-      List<ComparisonExclusionsConfiguration> configs = comparisonExclusionsConfigurationRepository.listBy(
-          appId, COMPARE_CONFIG_TYPE);
-      if (CollectionUtils.isNotEmpty(configs)) {
-        body.setComparisonExclusions(configs.stream().map(
-                ConfigComparisonExclusionsMapper.INSTANCE::voFromDto).collect(Collectors.toList()));
+      try {
+        CompareConfiguration compareConfiguration = queryConfigService.queryCompareConfiguration(appId);
+        body.setCompareConfiguration(compareConfiguration);
+      } catch (Exception e) {
+        LOGGER.error("query compare configuration error", e);
```
Comment (coryhh) [inline arex-storage-web-api/src/main/java/com/arextest/storage/web/controller/config/AgentRemoteConfigurationController.java:152]: the message need contain the info of "appId"

File: arex-storage-web-api/src/main/java/com/arextest/storage/service/QueryConfigService.java (hunk 10 above)
Comment (wildeslam) [reply arex-storage-web-api/src/main/java/com/arextest/storage/service/QueryConfigService.java:247]: to fetch the only element of a set

**Review by coryhh — APPROVED** [review]

**Review by lucas-myx — APPROVED** [review]

**Pull request merged**